server/src/main/webapp/WEB-INF/rails/webpack/views/pages/new-environments.tsx

```diff
@@ -19,23 +19,37 @@ import m from "mithril";
 import Stream from "mithril/stream";
 import {Environments, EnvironmentWithOrigin} from "models/new-environments/environments";
 import {EnvironmentsAPIs} from "models/new-environments/environments_apis";
+import * as Buttons from "views/components/buttons";
 import {FlashMessage, MessageType} from "views/components/flash_message";
+import {HeaderPanel} from "views/components/header_panel";
+import {CreateEnvModal} from "views/pages/new-environments/create_env_modal";
 import {EnvironmentsWidget} from "views/pages/new-environments/environments_widget";
 import {Page, PageState} from "views/pages/page";
+import {AddOperation, SaveOperation} from "views/pages/page_operations";
 
-interface State {
-  onSuccessfulSave: (msg: m.Children) => void;
+interface State extends AddOperation<EnvironmentWithOrigin>, SaveOperation {
 }
 
 export class NewEnvironmentsPage extends Page<null, State> {
   private readonly environments: Stream<Environments> = Stream(new Environments());
 
   oninit(vnode: m.Vnode<null, State>) {
     super.oninit(vnode);
+
+    vnode.state.onAdd = (e: MouseEvent) => {
+      e.stopPropagation();
+      this.flashMessage.clear();
```

Comment:
Avoid using `this` inside callback functions as a general practice.
Using `this` inside a callback function would require callee to bind `onAdd` method with an appropriate context object. 
This would mean doing `vnode.state.onAdd.bind(vnode.state)()` everytime we want to use this method.

In this example, `this` can be replaced with `vnode.state`.

Reply:
Got it. will update the code

Reply:
I forgot that this callback is an arrow function. And AFAIK, arrow functions bind to the context they are defined in. I am not sure when the issue that you mentioned will occur.

Reply:
> arrow functions bind to the context they are defined in

This is something new for me. Can you send me some documentation where I can read more about it?

Reply:
@vrushaliwaykole @dhanasp -- You are right.
With arrow functions, the `this` keyword always represents the object that defined the arrow function. In our case its the `vnode` object.

If we're using arrow function, can we remove the unnecessary `fn.bind(vnode)`?

Reply:
Yes. sure. I will look into it.